developers.md

```diff
@@ -0,0 +1,71 @@
+# For Developers
+
+Documentation and points of discussion for GreenX developers.
+
+## Portability
+
+* The minimax library has been tested on:
+  - macOS Catalina, with GCC9 and openBLAS
+  - Ubuntu-20.04, with GCC9 and blas/lapack
+  
+* There has been no testing with Intel or MKL. This should be done.
+
+## Testing 
+
+The test framework is currently set up such that one:
+* Writes a .f90 test driver, which calls some GreenX API
+* Writes a .py file to: 
+    * Supply inputs to the .f90 driver
+    * Execute the driver binary
+    * Parse the outputs of the test
+    * Make assertions
+    
+This has some requirements, and some technical open questions associated with it.
+Two better approaches would be to:  
+
+a). Replace a fortran unit testing framework. In most cases, this is more appropriate.
+     This would remove the need for all I/O and path-setting.
+     
+b). Call the respective fortran library directly from python, removing the need
+    for the .f90 driver, and removing all the complications associated with
+    consistent binary/file paths.
+
+### Issues with the Test Set Up
+
+#### Test Binary Name
+
+* The fortran binary name is specified in the python test.
+  - If one enforces the convention that the .f90 and .py files must have the same
+    prefix, the binary name can automatically be retrieved by python
+    `os.path.basename(__file__).py`
+
+#### Location of a Test Binary
+
+* The build directory root defined by python, from which to start searching for the 
+  test binary is not robust w.r.t change in directory nesting. As directory nesting 
+  is changed, one is forced to modify `root = Path(__file__).parent.parent` to 
+  `root = Path(__file__).parent.parent.parent`, etc.
+  
+* `find_test_binary` recursively searches for the binary, per test!
+   This will get inefficient very quickly, and should be viewed as a *place holder*.
+   - A prior implementation passed the full binary path to pytest, however
+   one must configure `conftest.py` such that it does not interfere with pytest's
+   existing command-line arguments. This is more robust than the current 
+   implementation. 
+      
+* Alex *assumes* relative paths will be a function of the directory in which pytest
+  is invoked. If one wants to to retain the flexbility of executing pytest in 
+  `<BUILD_DIR>`, `<BUILD_DIR>/test` or lower, relative binary paths are not a
+  good option. 
+  - One should investigate placing test binaries next to the python
+    test drivers, in the build folder. Perhaps relative paths can be used, 
+    removing the need to define `root` and `find_test_binary`. 
```

Comment:
So: Executables are compiled and go somewhere.  We need the root directory where those files reside.  At runtime, we can do a single glob under that root directory to find all the binaries, and then we can expose that as a fixture such that any test can name any file it wants.

The only dangling issue is how the python environment sees the binary root dir.  If it cannot be inferred from the build system, and the build system cannot provide any help (e.g. creating a symlink or setting an envvar), then it needs to be an argument to pytest, which however burdens the user with always providing that path.

Reply:
1. What performs the glob? As in, each test runs independently so where does one call/define the fixture for querying?

2. It has to be an argument. For CMake, one typically has multiple build dirs: Debug, Release, ReleaseWithDebug, some_build_configuration, arbitrary_name, etc.

Reply:
I don't have any very good answers, but it seems to me that *someone* (CMake or whoever it is) must know *where* things are, and be able to make sure that pytest knows about it e.g. in a session-scoped fixture.

Since I don't really know the build system and CMake directory stuff, I'll leave it to you to decide if/when anything should be changed.

Reply:
Yes, CMake knows the build directory. The issue is how to get it to pytest. I had a version last year that passed it as command line argument to pytest. 

One could have CMake set an ENV variable. Each time the developer builds in a new directory, it just gets overwritten with the updated variable. Sound reasonable?

I still don't understand how to execute point 1. What framework is performing the glob? For example if I had pytest do it, how do all tests get access to the fixture? Do it in the conftest file?

Reply:
Yes, if pytest does the glob inside a (e.g. session scoped) fixture in conftest.py, which is the standard place to put fixtures shared among the whole test suite and therefore appropriate for all "installation-level" info, then the tests can use the fixture like they use any other fixture.

The session-scoped fixture can hold *all* the executables somehow, and make them available as *names* somehow (e.g. in a dictionary).  The most direct way is just to return them as a dictionary.

Reply:
Done 👌